crates/transpiler/src/passes/optimize_clifford_t.rs

```diff
@@ -0,0 +1,591 @@
+                optimized_sequence.push(StandardGate::SXdg);
+                global_phase -= PI / 8.;
+            }
+            (Pauli1q::Y, true) => {
+                optimized_sequence.push(StandardGate::SX);
+                optimized_sequence.push(StandardGate::Tdg);
+                optimized_sequence.push(StandardGate::SXdg);
+                global_phase += PI / 8.;
+            }
+            (Pauli1q::Z, false) => {
+                optimized_sequence.push(StandardGate::T);
+                global_phase -= PI / 8.;
+            }
+            (Pauli1q::Z, true) => {
+                optimized_sequence.push(StandardGate::Tdg);
+                global_phase += PI / 8.;
+            }
+        }
+    }
+
+    // Insert the remaining Clifford gates
+    let (clifford_gates, phase_update) = clifford1q.to_circuit();
+    for gate in clifford_gates {
+        optimized_sequence.push(*gate);
+    }
+    global_phase += phase_update;
+
+    is_reduced.then_some((optimized_sequence, global_phase))
+}
+
+#[pyfunction]
+#[pyo3(name = "optimize_clifford_t")]
+pub fn run_optimize_clifford_t(dag: &mut DAGCircuit) -> PyResult<()> {
+    let op_counts = dag.get_op_counts();
+
+    // Stop the pass if there are unsupported gates.
+    if !op_counts
+        .keys()
+        .all(|k| CLIFFORD_T_GATE_NAMES.contains(&k.as_str()))
+    {
+        let unsupported: Vec<_> = op_counts
+            .keys()
+            .filter(|k| !CLIFFORD_T_GATE_NAMES.contains(&k.as_str()))
+            .collect();
+
+        return Err(TranspilerError::new_err(format!(
+            "Unable to run Clifford+T optimization as the circuit contains gates not supported by the pass: {:?}",
+            unsupported
+        )));
+    }
+
+    let runs: Vec<Vec<NodeIndex>> = dag.collect_1q_runs().unwrap().collect();
+
+    for raw_run in runs {
+        let optimized_sequence = optimize_clifford_t_1q(dag, &raw_run);
+        match optimized_sequence {
+            Some((optimized_sequence, global_phase_update)) => {
+                for gate in optimized_sequence {
+                    dag.insert_1q_on_incoming_qubit((gate, &[]), raw_run[0]);
+                }
+
+                dag.add_global_phase(&Param::Float(global_phase_update))?;
+                dag.remove_1q_sequence(&raw_run);
+            }
+            None => {
+                // No reductions were performed.
+            }
+        }
+    }
+    Ok(())
+}
+
+pub fn optimize_clifford_t_mod(m: &Bound<PyModule>) -> PyResult<()> {
+    m.add_wrapped(wrap_pyfunction!(run_optimize_clifford_t))?;
+    Ok(())
+}
+
+// Precomputed tables used in the algorithm.
+
+// Index of the Clifford1q operator -> corresponding Clifford circuit
+static CIRCUIT: [&[StandardGate]; 24] = [
+    &[],
+    &[StandardGate::H],
+    &[StandardGate::S],
+    &[StandardGate::H, StandardGate::S],
+    &[StandardGate::S, StandardGate::H],
+    &[StandardGate::S, StandardGate::H, StandardGate::S],
```

Comment:
A minor comment:
I think that for the 6 rep's for the Cliffords, it might be better to choose:
`[I, H, S, HS, SdgH, SHS]`
since they give exactly the following Cliffords:
```
Clifford: Stabilizer = ['+Z'], Destabilizer = ['+X']
Clifford: Stabilizer = ['+X'], Destabilizer = ['+Z']
Clifford: Stabilizer = ['+Z'], Destabilizer = ['+Y']
Clifford: Stabilizer = ['+Y'], Destabilizer = ['+Z']
Clifford: Stabilizer = ['+X'], Destabilizer = ['+Y']
Clifford: Stabilizer = ['+Y'], Destabilizer = ['+X']
```
You chose `SH` instead of `SdgH`, which gives "-" instead of "+":
`Clifford: Stabilizer = ['+X'], Destabilizer = ['-Y']`

Reply:
Perhaps we should add somewhere the code that generates these tables? 
Maybe as part of the tests? (in a seperate PR).